### ISSUE TITLE
Fix MLP Training Example Compilation

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#2706]
## Summary
The `examples/getting-started/mlp_training_example.mojo` has multiple compilation issues that need to be fixed.

## Issues to Fix
- [ ] Update `DynamicVector` import path (outdated Mojo stdlib path)
- [ ] Add missing `mean`/`mean_backward` exports to `shared.core/__init__.mojo`
- [ ] Replace `let` keyword with `var` for variable declarations
- [ ] Verify ExTensor implicit copy/move semantics
- [ ] Verify tuple return syntax

## Files
- `examples/getting-started/mlp_training_example.mojo`

## Related FIXMEs
- Line 37: `# FIXME: Check correct import path for DynamicVector`
- File header: `FIXME: This example has multiple compilation issues`

**Comment by mvillmow** [comment]
## Detailed Implementation Plan

Let me output the implementation plan directly as markdown:

---

# Implementation Plan: Fix MLP Training Example Compilation (#2706)

## Summary

The `examples/getting-started/mlp_training_example.mojo` file has multiple compilation issues
that prevent it from building. This plan addresses all identified issues:

1. **Missing `__init__.mojo`** in `getting-started` directory (module resolution fails)
2. **Undefined variables** (`input_size`, `hidden_size`, `output_size`, `X`, `y_true`)
3. **Tuple unpacking scope issue** - variables from tuple unpacking not visible in outer scope
4. **Shape initialization** - Empty List shapes create 0D scalars, not intended tensor sizes
5. **DynamicVector reference** - Outdated reference in FIXME comment (already using correct `List`)

**Note**: The original issue mentioned `let` keyword and `DynamicVector` imports, but the file
actually uses `var` correctly and doesn't import `DynamicVector` - it only mentions it in a
FIXME comment. The `mean`/`mean_backward` exports are already present in `shared.core/__init__.mojo` (lines 481, 485).

## Root Cause Analysis

### Issue 1: Module Resolution Failure

```text
error: unable to locate module 'shared'
from shared.core import (
     ^
```

The `examples/getting-started/` directory lacks an `__init__.mojo` file, which may affect module resolution when running from that directory.

**Fix**: Add `__init__.mojo` to `examples/getting-started/` directory.

### Issue 2: Undefined Variables

```text
error: use of unknown declaration 'hidden_size'
error: use of unknown declaration 'input_size'
error: use of unknown declaration 'output_size'
error: use of unknown declaration 'X'
error: use of unknown declaration 'y_true'
```

The code references `input_size`, `hidden_size`, `output_size` without defining them, and
`X`/`y_true` from tuple unpacking are not visible in the outer scope.

**Fix**: Add `alias` declarations for network sizes and fix tuple unpacking.

### Issue 3: Tuple Unpacking Scope

```mojo
var (X, y_true) = create_synthetic_data()
print("Input shape:", X.shape()[0], ...)  # ERROR: X unknown
```

In Mojo, tuple unpacking with `var (a, b) = ...` creates variables, but there may be scope
issues. The pattern works in other files (e.g., `shared/testing/data_generators.mojo:22`).

**Fix**: Use explicit unpacking:

```mojo
var data = create_synthetic_data()
var X = data.get[0, ExTensor]()
var y_true = data.get[1, ExTensor]()
```

### Issue 4: Empty Shape Lists

```mojo
var input_shape = List[Int]()  # Creates empty list
var inputs = ExTensor(input_shape, DType.float32)  # Creates 0D scalar
```

The code creates empty shape lists expecting to create tensors, but an empty shape creates
a 0D scalar tensor with 1 element, not the intended multi-element tensors.

**Fix**: Initialize shapes with actual dimensions:

```mojo
var input_shape = [4, 2]  # 4 samples, 2 features
var target_shape = [4, 1]  # 4 samples, 1 output
```

### Issue 5: FIXME Comment About DynamicVector

The FIXME comment mentions `DynamicVector` but the code doesn't actually import it. The
comment is misleading - the actual imports are correct using `List`.

**Fix**: Remove or update the FIXME comment.

## Step-by-Step Implementation Tasks

### Task 1: Create `__init__.mojo` in getting-started directory

**File**: `examples/getting-started/__init__.mojo`

**Content**:

```mojo
"""Getting started examples for ML Odyssey."""
```

**Rationale**: Establishes the directory as a Mojo package.

### Task 2: Define network architecture constants

**File**: `examples/getting-started/mlp_training_example.mojo`

**Location**: After line 135 (within `train_mlp()` function, after hyperparameters)

**Add**:

```mojo
    # Network architecture
    alias input_size = 2   # XOR has 2 input features
    alias hidden_size = 4  # Hidden layer size
    alias output_size = 1  # Binary classification
```

**Rationale**: Defines the missing constants used in shape construction.

### Task 3: Fix shape initialization in `create_synthetic_data()`

**File**: `examples/getting-started/mlp_training_example.mojo`

**Lines**: 82-86

**Change from**:

```mojo
    var input_shape = List[Int]()
    var target_shape = List[Int]()

    var inputs = ExTensor(input_shape, DType.float32)
    var targets = ExTensor(target_shape, DType.float32)
```

**Change to**:

```mojo
    # XOR data: 4 samples, 2 features each
    var input_shape = [4, 2]
    var target_shape = [4, 1]

    var inputs = ExTensor(input_shape, DType.float32)
    var targets = ExTensor(target_shape, DType.float32)
```

**Rationale**: Creates tensors with correct dimensions for XOR data.

### Task 4: Fix tuple unpacking in `train_mlp()`

**File**: `examples/getting-started/mlp_training_example.mojo`

**Lines**: 141-143

Use explicit unpacking:

```mojo
    # Create synthetic data
    print("\nCreating synthetic XOR data...")
    var data = create_synthetic_data()
    var X = data.get[0, ExTensor]()
    var y_true = data.get[1, ExTensor]()
    print("Input shape:", X.shape()[0], "x", X.shape()[1])
    print("Target shape:", y_true.shape()[0], "x", y_true.shape()[1])
```

**Rationale**: Ensures variables are accessible in outer scope.

### Task 5: Fix weight shape initialization

**File**: `examples/getting-started/mlp_training_example.mojo`

**Lines**: 149-158

**Change from**:

```mojo
    var W1_shape: List[Int] = [hidden_size, input_size]
    var b1_shape = List[Int]()
    var W1 = xavier_uniform(2, 4, W1_shape, DType.float32)
    var b1 = zeros(b1_shape, DType.float32)

    var W2_shape: List[Int] = [output_size, hidden_size]
    var b2_shape = List[Int]()
    var W2 = xavier_uniform(4, 1, W2_shape, DType.float32)
    var b2 = zeros(b2_shape, DType.float32)
```

**Change to**:

```mojo
    # Layer 1: (2, 4) - Input to Hidden
    var W1_shape = [hidden_size, input_size]  # [4, 2]
    var b1_shape = [hidden_size]  # [4]
    var W1 = xavier_uniform(input_size, hidden_size, W1_shape, DType.float32)
    var b1 = zeros(b1_shape, DType.float32)

    # Layer 2: (4, 1) - Hidden to Output
    var W2_shape = [output_size, hidden_size]  # [1, 4]
    var b2_shape = [output_size]  # [1]
    var W2 = xavier_uniform(hidden_size, output_size, W2_shape, DType.float32)
    var b2 = zeros(b2_shape, DType.float32)
```

**Rationale**: Correct shapes and fan_in/fan_out parameters for xavier initialization.

### Task 6: Fix sample tensor shapes

**File**: `examples/getting-started/mlp_training_example.mojo`

**Lines**: 176-179, 190-192, 200-201, 267-270

**Change** empty shape lists to proper dimensions:

```mojo
# Line 176-177: x_sample shape
var x_sample_shape = [input_size, 1]  # Column vector [2, 1]

# Line 190-191: y_sample shape
var y_sample_shape = [output_size, 1]  # Column vector [1, 1]

# Line 200-201: grad_loss shape
var grad_loss_shape = [1, 1]  # Scalar as 1x1 tensor

# Line 267-268: x_test shape (in testing loop)
var x_test_shape = [input_size, 1]  # Column vector [2, 1]
```

### Task 7: Update FIXME header comment

**File**: `examples/getting-started/mlp_training_example.mojo`

**Lines**: 15-35

**Remove or update** the FIXME block since most issues will be fixed. Change to:

```mojo
"""Complete 2-Layer MLP Training Example.

This example demonstrates end-to-end training of a 2-layer multi-layer perceptron
for binary classification using the newly implemented loss functions and optimizers.

Architecture:
    Input (2) -> Hidden (4) -> Output (1)
    Activations: ReLU (hidden), Sigmoid (output)
    Loss: Binary Cross-Entropy
    Optimizer: SGD

This example trains on synthetic XOR-like data to demonstrate that the training
pipeline is fully functional.
"""
```

### Task 8: Remove outdated FIXME comment

**File**: `examples/getting-started/mlp_training_example.mojo`

**Line**: 37

**Remove**:

```mojo
# FIXME(#2706): Check correct import path for DynamicVector
```

The imports are correct and don't use DynamicVector.

### Task 9: Fix add_backward tuple unpacking (Lines 216-232)

**File**: `examples/getting-started/mlp_training_example.mojo`

The `add_backward` and `matmul_backward` calls use tuple unpacking that may have scope issues.
Use explicit unpacking:

```mojo
        # Backprop through layer 2
        var add2_grads = add_backward(
            grad_z2, matmul(W2, h1).shape(), b2.shape()
        )
        var grad_h1_from_add = add2_grads.get[0, ExTensor]()
        var grad_b2 = add2_grads.get[1, ExTensor]()

        var matmul2_grads = matmul_backward(grad_h1_from_add, W2, h1)
        var grad_W2 = matmul2_grads.get[0, ExTensor]()
        var grad_h1 = matmul2_grads.get[1, ExTensor]()

        # Backprop through ReLU
        var grad_z1 = relu_backward(grad_h1, z1)

        # Backprop through layer 1
        var add1_grads = add_backward(
            grad_z1, matmul(W1, x_sample).shape(), b1.shape()
        )
        var grad_x_from_add = add1_grads.get[0, ExTensor]()
        var grad_b1 = add1_grads.get[1, ExTensor]()

        var matmul1_grads = matmul_backward(grad_x_from_add, W1, x_sample)
        var grad_W1 = matmul1_grads.get[0, ExTensor]()
        var grad_x = matmul1_grads.get[1, ExTensor]()
```

## Files to Modify/Create

| File | Action | Description |
|------|--------|-------------|
| `examples/getting-started/__init__.mojo` | Create | Package initialization |
| `examples/getting-started/mlp_training_example.mojo` | Modify | Fix all compilation issues |

## Testing Approach

### Step 1: Compilation Test

```bash
cd /home/mvillmow/ml-odyssey
pixi run mojo build examples/getting-started/mlp_training_example.mojo
```

**Expected**: Clean compilation with no errors.

### Step 2: Runtime Test

```bash
pixi run mojo run examples/getting-started/mlp_training_example.mojo
```

**Expected**: Program runs and shows training progress with loss decreasing.

### Step 3: Verify Training Convergence

Check output shows:

1. Loss starts high (> 0.5 for BCE)
2. Loss decreases over epochs
3. Predictions approach targets (0 or 1) after training
4. XOR pattern learned: [0,0]→0, [0,1]→1, [1,0]→1, [1,1]→0

### Step 4: Pre-commit Hooks

```bash
pre-commit run --files examples/getting-started/mlp_training_example.mojo
pre-commit run --files examples/getting-started/__init__.mojo
```

**Expected**: All hooks pass (formatting, linting).

## Success Criteria

- [ ] `mojo build` completes without errors
- [ ] `mojo run` executes successfully
- [ ] Training loop runs for 1000 epochs without crashes
- [ ] Loss decreases during training (indicates learning)
- [ ] Test predictions show reasonable approximation of XOR function
- [ ] Pre-commit hooks pass
- [ ] FIXME comments removed or updated

## Risk Assessment

| Risk | Likelihood | Mitigation |
|------|------------|------------|
| Tuple unpacking scope issues | Medium | Use explicit unpacking as fallback |
| Numerical instability | Low | Small learning rate, stable activations |
| Shape mismatches in backward | Medium | Verify each backward pass shape |
| Memory safety | Low | ExTensor handles ownership correctly |

## Dependencies

- No external dependencies
- Uses existing `shared.core` and `shared.training.optimizers` modules
- `mean` and `mean_backward` already exported from `shared.core/__init__.mojo` (lines 481, 485)

## Resource Usage

- **Tool calls used**: 24
- **Steps taken**: 16
- **Files read**: 12
- **Searches performed**: 10
- **Remaining budget**: 26 tool calls, 34 steps
